compiler/sem/collectors.nim

```diff
@@ -46,18 +99,310 @@ type
+      processVarSection(n, structs)
+    of nkStmtList:
+      unreachable("top-level statement lists must not exist at this point")
+    else:
+      scan(n, structs)
+
+  for it in stmts.items:
+    process(it, structs)
+
+
+proc generateModuleDestructor(graph: ModuleGraph, m: Module): PNode =
+  ## Generates the body for the destructor procedure of module `m` (also
+  ## referred to as the 'de-init' procedure).
+  result = newNode(nkStmtList)
+  for i in countdown(m.structs.globals.high, 0):
+    let s = m.structs.globals[i]
+    if hasDestructor(s.typ):
+      result.add genDestroy(graph, m.idgen, m.sym, newSymNode(s))
+
+  # note: the generated body is not yet final -- we're still missing
+  # destructor calls for globals defined inside procedures (i.e., pure
+  # globals). These calls are inserted once we know all alive pure globals
+  # (which is after the main part of code generation has finished)
+
+  if result.len == 0:
+    # collapse to an empty node (dead-code elimination looks for them in
+    # order to detect empty procedures)
+    result = newNode(nkEmpty)
+
+proc changeOwner(n: PNode, newOwner: PSym) =
+  ## For all symbols defined in the AST `n`, changes the owner to
+  ## `newOwner`.
+  template change(it: PNode) =
+    # make sure to not change the owner of globals
+    # XXX: we shouldn't need to check for globals here, but have to,
+    #      because of globals defined in nested scopes
+    if it.kind == nkSym and sfGlobal notin it.sym.flags:
+      it.sym.owner = newOwner
+
+  case n.kind
+  of nkIdentDefs, nkVarTuple, nkConstDef:
+    for it in names(n):
+      change(it)
+
+    # ignore the type slot here. If it contains definitions, so be
+    # it -- changing their owner shouldn't cause any problems
+    changeOwner(n[^1], newOwner)
+  of routineDefs - nkIteratorDef:
+    change(n[namePos])
+  of nkLambdaKinds:
+    change(n[namePos])
+  of nkForStmt:
+    # 'for' statements don't use identdefs...
+    for it in names(n):
+      case it.kind
+      of nkSym:
+        change(it)
+      of nkVarTuple:
+        # not a normal var tuple...
+        for i in 0..<it.len-1:
+          change(it[i])
+      else:
+        unreachable()
+
+    changeOwner(n[^2], newOwner)
+    changeOwner(n[^1], newOwner)
+  of nkIteratorDef:
+    # we special case iterator definitions here, and leave them as direct
+    # descendants of the module. This is currently done for efficieny
+    # reasons: without, a top-level closure iterator creating an instance of
+    # another closure iterator would lead to the unnecessary allocation of
+    # an empty environment that lives for the duration of the module-init
+    # procedure.
+    # XXX: ^^ this is limitation of the lambda-lifting pass
+    discard
+  of nkWithoutSons, nkTypeSection:
+    discard "ignore"
+  of nkWithSons - entityDefs - nkTypeSection:
+    for it in n.items:
+      changeOwner(it, newOwner)
+
+proc setupModule*(graph: ModuleGraph, idgen: IdGenerator, m: PSym,
```

Comment:
The procedure might need a more descriptive name. While it does set up a `Module` instance, it also does a lot more.